x/programs/cmd/simulator/src/lib.rs

```diff
@@ -174,12 +148,12 @@ pub struct PlanResult {
 }
 
 #[derive(Debug, Deserialize)]
-pub struct PlanResultTyped<T>
+pub struct StepResultTyped<T>
```

Comment:
I'd have loved to, but it's not easy to avoid having an intermediary type (if possible?).
The issue is that the return type has to somehow be known beforehand.
- `impl BorshDeserialize` would be great, but it's not allowed in fields. Otherwise I would have put it in `StepReponse`
- `Box<dyn BorshDeseialize>` to do the dispatching at runtime, but this type doesn't implement `Deserialize` which makes sense

Reply:
FYI

Neither `impl BorshDeserialize` nor `Box<dyn BorshDeserialize>` wouldn't make any sense here. `impl Trait` syntax is meant for opaque types that are known at compile-time OR they are equivalent to a generic with a trait bound. `Box<dyn Trait>` requires object-safety ([refresher](https://doc.rust-lang.org/reference/items/traits.html#object-safety)). `BorshDeserialize` is not object-safe.

Reply:
Why don't you just remove the `pub` visibility from `StepResult::response` and have a getter method. 
```rust
impl StepResult {
    fn response<T: BorshDeserialize>(&self) -> Result<T, borsh::io::Error> {
        borsh::from_slice(&self.response)
    } 
}
```
Then you can get rid of all the `*Typed<T>` types

Reply:
Thanks for the refresher, and for the very nice solution. It solves it all...